experimental/transliterate/src/compile/mod.rs

```diff
@@ -39,43 +38,122 @@ impl Direction {
+///       {$OE} [:Lowercase:] → Oe;
+///       {$UE} [:Lowercase:] → Ue;
+///  
+///       $AE → AE;
+///       $OE → OE;
+///       $UE → UE;
+///  
+///       ::Any-ASCII;
+///     "#.to_string(),
+///     // legacy alias, doesn't actually do anything in this case
+///     ["de-ASCII"],
+///     false,
+///     true,
+/// );
+/// collection.register_source(
+///     &"und-t-und-latn-d0-ascii".parse().unwrap(),
+///     "# ... ".to_string(),
+///     // the previous transliterator refers to this one by the `Any-ASCII` ID,
+///     // so it's important that the alias is listed here.
+///     ["Latin-Ascii", "Any-ASCII"],
+///     false,
+///     true,
+/// );
+///
+/// Transliterator::try_new_unstable("de-t-de-d0-ascii".parse().unwrap(), &collection.as_provider()).unwrap();
+///
+pub struct RuleCollection {
+    id_mapping: HashMap<String, Locale>, // alias -> bcp id
+    data: RwLock<HashMap<DataLocale, (String, bool, bool)>>, // locale -> source/reverse/visible
+    cache: RwLock<HashMap<DataLocale, DataResponse<TransliteratorRulesV1Marker>>>,
 }
 
-impl RuleBasedTransliterator<'static> {
-    /// Parse a rule based transliterator definition into a `TransliteratorDataStruct`.
-    ///
-    /// See [UTS #35 - Transliterators](https://unicode.org/reports/tr35/tr35-general.html#Transforms) for more information.
-    ///
-    /// The `direction` [`Direction`] argument
-    /// determines which of the returned options is populated. The first option will be populated
-    /// if the direction is [`Forward`](Direction::Forward), the second if the direction is
-    /// [`Reverse`](Direction::Reverse), and both if the direction is [`Both`](Direction::Both).
+impl RuleCollection {
+    /// Add a new transliteration source to the collection.
+    pub fn register_source<'a>(
+        &mut self,
+        id: &icu_locid::Locale,
+        source: String,
+        aliases: impl IntoIterator<Item = &'a str>,
+        reverse: bool,
+        visible: bool,
+    ) {
+        #[allow(clippy::expect_used)]
+        self.data.get_mut().expect("poison").insert(
+            crate::ids::bcp47_to_data_locale(id),
+            (source, reverse, visible),
+        );
+
+        for alias in aliases.into_iter() {
+            self.id_mapping
+                .insert(alias.to_ascii_lowercase(), id.clone());
+        }
+    }
+
+    /// Returns a provider usable by [`Transliterator::try_new_unstable`](crate::Transliterator::try_new_unstable).
     ///
-    /// `id_mapping` is a map from legacy IDs to internal ICU4X IDs. Occurring IDs in `source`
-    /// that do not have a corresponding entry in `id_mapping` are still valid, but will be
-    /// compiled with [`legacy_id_to_internal_id`].
-    pub fn compile<P>(
-        provider: &P,
-        source: &str,
-        direction: Direction,
-        id_mapping: &HashMap<String, String>,
-    ) -> Result<(Option<Self>, Option<Self>)>
+    /// ✨ *Enabled with the `compiled_data` Cargo feature.*
```

Comment:
Question: do we need the stability warning for the `impl Trait` return value bounds?